implementation/leprechaun/src/shared/api/filter.ts

```diff
@@ -0,0 +1,28 @@
+import { Type } from '@nestjs/common';
+import { ApiQueryOptions } from '@nestjs/swagger';
+import { plainToInstance } from 'class-transformer';
+import { validateSync } from 'class-validator';
+
+// TODO: there is a way to force TypeScript to check if the array of objects matches EXACTLY the interface
+// source? came up in a dream
+export type ForceArrayToHaveNamedApiQueryOf<T extends {}> = Array<{ name: keyof T } & ApiQueryOptions>;
+
+// NOTE: Query params should be soft-validated, if they are ok then they are passed to the decorator, otherwise no
+//       exception should be thrown, and offending fields should be removed. class-validator alone is not enough
+//       thus the removeOffendingFields function
+export function removeOffendingFields<T extends {}>(data: T, type: Type<T>): T {
+  const obj = plainToInstance(type, data);
+  const copy = { ...data };
+  validateSync(obj, {
+    whitelist: true, // remove unknown properties
+    skipMissingProperties: true, // allow missing properties
+    forbidUnknownValues: true, // remove unknown values
+    forbidNonWhitelisted: true, // remove non-whitelisted properties
+    stopAtFirstError: false, // continue validation after first error
+  }).map((error) => delete copy[error.property as keyof T]);
+  return copy;
+}
+
+// https://www.30secondsofcode.org/js/s/escape-reg-exp/
+export const escapeRegExp = (str: string) => str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+export const EscapedRegExpLike = (str: string) => RegExp(escapeRegExp(str));
```

Comment:
Sounds like a security AND performance nightmare. Can't we do something [like this](https://mikro-orm.io/docs/query-conditions#postgresql)?

Reply:
The issue with fulltext search here is that none of the values of `line.names`, `stop.names`, `sideNumbers` would make any sense to fulltext indexer as it depends on some structure of indexed text. Changing table to be used with indexes (as per your link) would brake `VehicleService`, and breaks some DTOs. Also, fulltext search is weird, it does not return values you'd hope for.